Add governance doc

Elchi3 opened the pull request:
Rendered version https://github.com/mdn/browser-compat-data/blob/ac7abfe89485764f6acf30be5f5dcacd3c1f3b0c/governance.md

This document aims to clarify how this project is governed and how we aim to give out permissions and with what privileges and responsibilities these permissions will come. For now, I've added the current peers, but the goal is to have more peers so that more people can review and merge the high amount of PRs this project is seeing.

ddbeck commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+
+- Manage and merge non-data pull requests. Such as schema, linter or infrastructure changes.
+- May merge their own pull requests once they have collected the feedback they deem necessary. (No pull request should be merged without at least one peer or owner comment stating they’ve looked at the PR.)
+- Release a new npm version of the project and a regular (weekly) basis.
```
```suggestion
- Release a new npm version of the project on a regular (weekly) basis.
```

ddbeck commented:
I'm really excited by this governance doc, Florian. I really like that you're putting some thought into broadening community involvement in the project. I also like that it's derived from other projects' governance docs, not invented from scratch.

One area that caught my attention is that owners are responsible for doing releases. This makes sense to me, but as someone who has done a few releases for you in the past, I thought it might be helpful, as a practical matter, to allow for some kind of owner-delegate rule (riffing a bit on Python's old "[BDFL-Delegate](https://www.python.org/dev/peps/pep-0001/#python-s-bdfl)" process). Something like this, maybe:

> Owners may, at their discretion, nominate an owner-delegate to carry out a task or make a decision ordinarily carried out by an Owner. Delegation should be limited in duration or scope, or both; delegation may be withdrawn by any Owner at any time. For example, an Owner may nominate an owner-delegate to approve an infrastructure PR or to publish npm packages for a set time or number of releases.

I think this might satisfy some practical needs of the project, like doing releases, or bringing in someone to help make a decision on an infrastructure change. For instance, this might've been a good step to do in #3004—instead I took a leap, which would've been (rightly) out-of-bounds under this governance doc.

I don't have my heart set on this being included, but I thought it might formalize something we've been doing haphazardly. Or I can just stop doing releases while you're on holiday. Whichever. 😆

Elchi3 commented:
Note to self: When granting people commit permissions, we probably also need to mention https://www.mozilla.org/en-US/about/governance/policies/commit/requirements/ in there and somehow record this statement from future peers: "I have read, and agree to abide by, the Commit Access Requirements.” 
See also the "Becoming A Mozilla Committer" procedure here https://www.mozilla.org/en-US/about/governance/policies/commit/

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+Everyone who is involved in any form with the project must abide by the project’s [Contribution Guidelines](https://github.com/mdn/browser-compat-data/blob/master/CODE_OF_CONDUCT.md). Everyone is expected to be respectful of every community member and to work collaboratively in the spirit of inclusion.
+
+### Users
+Users are community members who have a need for the project. They are typically consumers of the compat data. Anyone can be a User; there are no special requirements and the data is licensed under [CC0](https://github.com/mdn/browser-compat-data/blob/master/LICENSE). Common User contributions include evangelizing the project (e.g., display a link on a website and raise awareness through word-of-m ...
```
Ah, I see you are using "User", "Contributor", etc. as proper nouns, e.g. job titles. I am cool with that, so feel free to disregard this comment.

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+
+Contributors may also review pull requests. This can be helpful, but their approval or disapproval is not decisive for merging or not merging PRs.
+
+As Contributors gain experience and familiarity with the project, their profile within, and commitment to, the community will increase. At some stage, they may find themselves being nominated for committership by an existing Peer or Owner.
```
Define somewhere what "committership" means.

Elchi3 replied:
I removed this term and replaced it with Peers throughout as it actually means Peers and is not just about committing.

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+
+Users who continue to engage with the project and its community will often become more and more involved. Such Users may find themselves becoming [Contributors](#Contributors), as described in the next section.
+
+### Contributors
```
Peers and Owners have an easy to follow bulleted list of what they can do. Why not represent the things Contributors can do in this format too, to keep things consistent?

Elchi3 replied:
Good idea! Done.

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+- Are expected to work on public branches of their forks and submit pull requests to the master branch.
+- Must submit pull requests for all their changes.
+- May label and close issues.
+- May merge compat data update pull requests only.
```
It took me a couple of scans to understand this correctly; I thought there was a missing word initially. How about "May only merge pull requests relating to compat data updates."

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+- Must submit pull requests for all their changes.
+- May label and close issues.
+- May merge compat data update pull requests only.
+- Have their non-data update work reviewed by [Owners](#Owners) before acceptance into the repository. Non-data pull requests are PRs that change the schema, update project meta-docs, the linter. or other infrastructure changes.
```
This bullet implies that they can merge compat data PRs that they've written without needing a review, when taken in context with the previous bullet. This isn't what we want surely. The previous bullet should make it clear that we are talking about them merging other people's PRs, and this one should say that their data update PRs need reviewing by another contributor or owner? Or put that in a separate bullet?

Elchi3 replied:
I think it is more clear now, but maybe have a second look at it.

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+
+To become a Peer:
+
+- One must have shown a willingness and ability to participate in the project as a team player. Typically, a potential Peer will need to show that they have an understanding of and alignment with the project, its objectives, and its strategy.
```
"one must" should go into the previous paragraph, so that the bullet structures are consistent and make sense.

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+
+New Peers can be nominated by any existing Peers. Once they have been nominated, there will be a vote by the Owners.
+
+It is important to recognize that being a Peer is a privilege, not a right. That privilege must be earned and once earned it can be removed by the Owners. However, under normal circumstances committership exists for as long as the Peer wishes to continue engaging with the project. Inactive Peers might be marked as inactive or removed by the Owners and may re-enter when they choose to contribute a ...
```
"removed by the Owners" -> "removed by the Owners if they violate our code of conduct or do not follow the rules listed above" ?

Elchi3 replied:
This is in the context of inactivity.

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+
+New Peers can be nominated by any existing Peers. Once they have been nominated, there will be a vote by the Owners.
+
+It is important to recognize that being a Peer is a privilege, not a right. That privilege must be earned and once earned it can be removed by the Owners. However, under normal circumstances committership exists for as long as the Peer wishes to continue engaging with the project. Inactive Peers might be marked as inactive or removed by the Owners and may re-enter when they choose to contribute a ...
```
"Inactive Peers might be marked as inactive or removed" -> Does it make sense to define somewhere what kind of time period we are talking about? No activity on the project for longer for a few months or more?

Elchi3 replied:
Added, good idea!

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+- Estelle Weyl (@estelle), Mozilla
+- John Whitlock (@jwhitlock), Mozilla
+
+Some Peers are also [MDN Content Curators](https://developer.mozilla.org/en-US/docs/MDN/Contribute/Documentation_topics_and_curators) and have thus earned the privilege to be a mdn-browser-compat-data peer, so that their expertise in a given content area (like CSS or HTML, etc.) can help improve the compat data for that same content area. Such peers are marked in the relevant folders using GitHub ...
```
"like CSS or HTML, etc." -> "CSS, HTML, etc."

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+- Estelle Weyl (@estelle), Mozilla
+- John Whitlock (@jwhitlock), Mozilla
+
+Some Peers are also [MDN Content Curators](https://developer.mozilla.org/en-US/docs/MDN/Contribute/Documentation_topics_and_curators) and have thus earned the privilege to be a mdn-browser-compat-data peer, so that their expertise in a given content area (like CSS or HTML, etc.) can help improve the compat data for that same content area. Such peers are marked in the relevant folders using GitHub ...
```
This paragraph and the following one could also be true of Owners, right? Is it worth saying "Peers and Owners" instead of just "Peers" in both cases, and moving them to a separate section after the "Peers" and "Owners" descriptions called something like "Additional paths to becoming a Peer or Owner"?

Elchi3 replied:
I've moved this into its own section. Good suggestion, thank you! :)

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+- GitHub repository hosting.
+
+Being an Owner is not time-limited. There is no fixed size of the Owner group. The Owner group should be of such a size as to ensure adequate coverage of important areas of expertise balanced with the ability to make decisions efficiently.
+An Owner may be removed from the Owner group by voluntary resignation, or by a standard Owner group motion.
```
"standard Owner group motion" -> This could use a definition. Also, isn't remaining an Owner also subject to the same scrutiny as remaining a Peer, e.g. following the rules and not violating the code of conduct?

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+
+Owners fulfill all requirements of Peers, and also:
+
+- Manage and merge non-data pull requests. Such as schema, linter or infrastructure changes.
```
"requests. Such as schema, linter or infrastructure changes." -> "requests such as schema, linter, or infrastructure changes."

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+- May merge their own pull requests once they have collected the feedback they deem necessary. (No pull request should be merged without at least one peer or owner comment stating they’ve looked at the PR.)
+- Release a new npm version of the project on a regular (weekly) basis.
+
+To become an Owner one must fulfill at least the following items and commit to being a part of the community for the long-term.
```
"items" -> "conditions"

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+
+To become an Owner one must fulfill at least the following items and commit to being a part of the community for the long-term.
+
+- Work in a helpful and collaborative way with the MDN community.
```
"Work in a helpful" -> "Have worked in a helpful"

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+
+- Work in a helpful and collaborative way with the MDN community.
+- Have given good feedback on others’ submissions and displayed an overall understanding of the code quality standards for the project.
+- Ability to drive the project forward, manage requirements from users, and taking on responsibility for the overall health of the project.
```
"Ability" -> "Have the ability"

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+The owners are also listed as top-level owners in GitHub’s code owner file.
+
+## Project Meetings
+There are no recurrent project meetings, they are scheduled when required, at a time that works for the owners, and using tools that enable participation by the community. The meeting is run by a designated moderator approved by the owners group.
```
"meetings, they are scheduled when required, at a time that works for the owners," -> "meetings; they are scheduled when required at a time that works for the owners,"

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+There are no recurrent project meetings, they are scheduled when required, at a time that works for the owners, and using tools that enable participation by the community. The meeting is run by a designated moderator approved by the owners group.
+Items are added to the agenda which are considered contentious or are modifications of governance, contribution policy, owner membership, or release process.
+
+The intention of the agenda is not to approve or review all patches. That should happen continuously on GitHub and be handled by the larger group of Peers. The exception to this is when defining how the schema should look like (or when proposing an update), there is not an initial agreement in the GitHub discussion, and action needs to be taken.
```
"look like" -> "look"

chrisdavidmills replied:
"update), there is not an initial agreement in the GitHub discussion, and action needs to be taken." -> I am a bit confused by the structure here. Should it be something like "update), or when a PR discussion has stalled due to disagreement or inaction, and progress needs to be unblocked."

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+The moderator is responsible for summarizing the discussion of each agenda item and adding it to the repository's wiki after the meeting.
+
+### Consensus Seeking Process
+Owners follow a [Consensus-seeking decision-making](https://en.wikipedia.org/wiki/Consensus-seeking_decision-making) model.
```
"Owners" -> should this be "Meetings"? It is down to everyone participating in a meeting, surely, not just the owners?

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+
+When an agenda item has appeared to reach a consensus, the moderator will ask “Does anyone object?” as a final call for dissent from the consensus.
+
+If an agenda item cannot reach a consensus, an owner  can call for either a closing vote or a vote to table the issue to the next meeting. The call for a vote must be approved by a majority of the owners or else the discussion will continue. Simple majority wins.
```
"owner  can" -> extra space

chrisdavidmills commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+### Consensus Seeking Process
+Owners follow a [Consensus-seeking decision-making](https://en.wikipedia.org/wiki/Consensus-seeking_decision-making) model.
+
+When an agenda item has appeared to reach a consensus, the moderator will ask “Does anyone object?” as a final call for dissent from the consensus.
```
I'm having a hard time with this bit — this sentence implies that any dissent will stop a motion going forward, and require discussions to resume. But then the following paragraph says that if a consensus (which is basically a majority in this context?) cannot be reached, a vote can be called for to close the issue. Does this mean close it without action, or vote what solution to the issue should be moved forward with? In which case, this is a call for a majority anyway?

Surely we should keep discussing the issue until all points of view have been heard, once that is done put the options for a way forward on the table, and then vote for which option is best in the case of a disagreement? Is this what the text is supposed to be saying anyway? If it is, it is a bit of a confusing way to say it.

Elchi3 commented on governance.md:
```diff
@@ -0,0 +1,137 @@
+Everyone who is involved in any form with the project must abide by the project’s [Contribution Guidelines](https://github.com/mdn/browser-compat-data/blob/master/CODE_OF_CONDUCT.md). Everyone is expected to be respectful of every community member and to work collaboratively in the spirit of inclusion.
+
+### Users
+Users are community members who have a need for the project. They are typically consumers of the compat data. Anyone can be a User; there are no special requirements and the data is licensed under [CC0](https://github.com/mdn/browser-compat-data/blob/master/LICENSE). Common User contributions include evangelizing the project (e.g., display a link on a website and raise awareness through word-of-m ...
```
Added a link to the projects using the data.

Elchi3 commented:
>  thought it might be helpful, as a practical matter, to allow for some kind of owner-delegate rule

Thanks @ddbeck! I've now added this as a section. Great idea!

> Note to self: When granting people commit permissions, we probably also need to mention https://www.mozilla.org/en-US/about/governance/policies/commit/requirements/ in there and somehow record this statement from future peers: "I have read, and agree to abide by, the Commit Access Requirements.”
See also the "Becoming A Mozilla Committer" procedure here https://www.mozilla.org/en-US/about/governance/policies/commit/

I have now added this to "Becoming a Peer", but I still need to double check what the standard Mozilla procedure is for contributor access to GitHub repos.

> Nitpicky review coming up...sorry about this ;-)

Thanks for your review. I have addressed many of your points now, but I left a few unresolved as I still need to get to them.

>   We were discussing possibly adding something here about priority by which we should try to review PRs. Did you think this should not go here? I'm happy either way really.
> Also, should we include here something about process for reviewing a PR, e.g. making sure you call for a review by at least one Peer/Owner, etc. Or again, is this not suitable here? I'm just trying to make sure we don't lose anything.

I'm not sure yet where or if this fits into this document. I think Peers and Owners will have their own review style and as long as the project runs healthy overall, it should be fine. PRs should have clear intents and standard pings for (re-)requesting review or or gentle nudging is standard practice. Maybe such guidelines fit better into the contributing doc?

chrisdavidmills commented:
> Thanks for your review. I have addressed many of your points now, but I left a few unresolved as I still need to get to them.

This is fine. There were a lot of points here, and what you've done in response massively improves the document. I think it is OK to iterate after publication.

> I'm not sure yet where or if this fits into this document. I think Peers and Owners will have their own review style and as long as the project runs healthy overall, it should be fine. PRs should have clear intents and standard pings for (re-)requesting review or or gentle nudging is standard practice. Maybe such guidelines fit better into the contributing doc?

I'll leave it in your hands to consider adding somewhere later if we feel it is needed.

wbamberg commented on governance.md:
```diff
@@ -0,0 +1,145 @@
+
+To become an Owner one must fulfill at least the following conditions and commit to being a part of the community for the long-term.
+
+- Have worked in a helpful and collaborative way with the MDN community.
```
This seems very like "team player" from https://github.com/mdn/browser-compat-data/pull/3668/files#diff-898e7f1ceacb493c024554f5a7c87bdfR39, and I like this wording better than "team player".

I like the idea of stressing "helpfulness to contributors" as a desirable attribute of both peers and owners. A big part of the role is interacting with people who have much less familiarity with the project and working with them to get their contributions merged.

wbamberg commented on governance.md:
```diff
@@ -0,0 +1,145 @@
+As Contributors gain experience and familiarity with the project, their profile within, and commitment to, the community will increase. At some stage, they may find themselves being nominated for becoming a Peer by an existing Peer or Owner.
+
+### Peers
+Peers are community members who have shown that they are committed to the continued development of the project through ongoing engagement with the community. Peers are given push/write access to the project’s GitHub repos.
```
Is it an _expectation_ that peers should review and merge data PRs? I mean this doc says that they _can_, but is signing up to some of that workload an expectation?

jpmedley replied:
This is a valid question. I help out whenever I can, but MDN doesn't pay my salary.

chrisdavidmills replied:
I think being a peer brings wiith it some expectation that you will do some data reviewing/merging, otherwise there is not much point you having the access privs you are given. If you don't do many PR reviews/merges, I guess it would still be acceptable peer duties to do other things like contribute to strategy dicussions and help with updates to the underlying code/other structures.

It is something we'll have to use our judgement on — if someone is contributing constructively, then fine.

Joe, I think the amount of input you've made definitely makes you elligable for consideration as a peer. I think many/most of the contributors to the project will end up being volunteers, not paid by MDN, and we will be mindful of that.

Elchi3 replied:
As Chris says: There is an expectation that peers do something with their peer status because otherwise there is no need for them to have these access privileges. However, there is no expectation like "must review 10 PRs a week" or similar. The hope is that we will share workload more among the peers, but no one will require from anyone to get a specific amount of work done.

chrisdavidmills commented:
@jpmedley see the section of the doc following "To become a Peer one must"

I am happy to nominate you.

jwhitlock commented on governance.md:
```diff
@@ -0,0 +1,145 @@
+- Have shown a willingness and ability to participate in the project as a team player. Typically, a potential Peer will need to show that they have an understanding of and alignment with the project, its objectives, and its strategy.
+- Have contributed a significant amount of work to the project (e.g. in the form of PRs or PR reviews), thereby demonstrating their trustworthiness and commitment to the project.
+- Read and agree to abide by the [Mozilla Commit Access Requirements](https://www.mozilla.org/en-US/about/governance/policies/commit/requirements/).
+- File an issue in the [mdn/mdn](https://github.com/mdn/mdn) repository and record “I have read, and agree to abide by, the Mozilla Commit Access Requirements.”
```
An alternate would be to have an issue in this repo, or even skip the issue and just make it a PR. For the ADRs, we made "approval" a commit against the PR (see the history for [ADR 002](https://github.com/mdn/mdn/commits/master/ADRs/002-use-flow-for-typechecking.md). In a similar way, a user could add their name to the "List of peers" below as a PR, with the required statement as the PR description or commit comment. This would make ``git blame`` useful for cross-referencing.

But, mdn/mdn is fine as well. It is easy to link to an issue in another repo.

Elchi3 replied:
I believe mdn/mdn is a good place for this. I also think that we would only require to record this statement once in case people apply for being a peer on multiple repos. Maybe mdn/mdn is a good place to keep record of all peers and owners and applications throughout our org. I hope to use it more as MDN makes its way to GitHub.

jwhitlock commented on governance.md:
```diff
@@ -0,0 +1,145 @@
+- Eric Shepherd (@a2sheppy), Mozilla
+- Irene Smith (@irenesmith), Mozilla
+- Estelle Weyl (@estelle), Mozilla
+- John Whitlock (@jwhitlock), Mozilla
```
Thanks for thinking of me. I'm probably a good candidate for the "removing an inactive peer" process 🙈

jwhitlock commented on governance.md:
```diff
@@ -0,0 +1,145 @@
+
+The intention of the agenda is not to approve or review all patches. That should happen continuously on GitHub and be handled by the larger group of Peers. The exception to this is when defining how the schema should look (or when proposing an update), or when a PR discussion has stalled due to disagreement or inaction, and progress needs to be unblocked.
+
+Any community member or Peer can ask that something be added to the next meeting’s agenda by logging a GitHub Issue. Peers can add the item to the agenda by adding the “meeting agenda” tag to the issue and Contributors can ask Peers to add the label for them.
```
"meeting agenda" should be a link to the tag, like [meeting-agenda](https://github.com/mdn/browser-compat-data/labels/meeting-agenda). I think you should go ahead and create the desired tag.

jwhitlock commented on governance.md:
```diff
@@ -0,0 +1,145 @@
+
+The Owners may invite persons or representatives from certain projects to participate in a non-voting capacity.
+
+The moderator is responsible for summarizing the discussion of each agenda item and adding it to the repository's wiki after the meeting.
```
I'd like a link to the agenda is needed. I'd suggest https://github.com/mdn/browser-compat-data/wiki/Weekly-meeting-notes. I'd also suggest making that page the wiki homepage.

Elchi3 commented:
Thanks for your feedback, John! I've added the links you've suggested and also removed you as a peer. Thanks for everything you have done for BCD and thank you specifically for the predecessor project which was inspiring and helpful for BCD to take off.

Elchi3 commented:
This document was also reviewed by members of Mozilla's Open Innovation team (https://wiki.mozilla.org/Innovation#The_Open_Innovation_Team). I want to thank them and all previous reviewers (members of the MDN team and members of the MDN Product Advisory Board [PAB]) as well as everyone who provided review in this PR.
